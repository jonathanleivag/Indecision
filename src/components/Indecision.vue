<template>
  <img v-if="image" :src="image" alt="bg" />
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input v-model="question" type="text" placeholder="Hazme una pregunta" />
    <p>Recuerda de terminar la pregunta con un signo de interrogaci√≥n (?)</p>
    <div v-if="isValidQuestion">
      <h2>{{ question }}</h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  name: 'indecision',
  data () {
    return {
      question: null,
      answer: null,
      image: null,
      isValidQuestion: false
    }
  },
  methods: {
    async getAnswer () {
      this.answer = 'pensando...'
      try {
        const { answer, image } = await window
        .fetch('https://yesno.wtf/api')
        .then(response => response.json())
        const resSpanish = { yes: 'Si', no: 'No', maybe: 'Puede ser' }
        this.answer = resSpanish[answer]
        this.image = image
      } catch (e) {
        this.answer = 'error'
      }
    }
  },
  watch: {
    question (value) {
      console.log({value});
      if (!value.includes('?')) {
        this.isValidQuestion = false
        this.image = null
        this.answer = null
      } else {
        this.isValidQuestion = true
        this.getAnswer()
      }
    }
  }
}
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 10px;
  margin-top: 10px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>
