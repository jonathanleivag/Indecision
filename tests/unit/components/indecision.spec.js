import { shallowMount } from '@vue/test-utils'
import Indecision from '@/components/Indecision'

describe('Indecision component', () => {
  const answer = 'yes'
  const forced = false
  const image = 'https://yesno.wtf/assets/yes/2.gif'

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer,
          forced,
          image
        })
    })
  )

  it('should match the snapshot', async () => {
    const wrapper = shallowMount(Indecision).html()
    expect(wrapper).toMatchSnapshot()
  })
  it('nothing should be executed when typing', async () => {
    const wrapper = shallowMount(Indecision)
    const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
    const input = wrapper.find('input')
    const clsSpy = jest.spyOn(console, 'log')
    await input.setValue('Hola mundo')
    expect(clsSpy).toHaveBeenCalled()
    expect(getAnswerSpy).not.toHaveBeenCalled()
  })

  it('The writing "?" I should run the getAnswer', async () => {
    const wrapper = shallowMount(Indecision)
    const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
    const input = wrapper.find('input')
    const clsSpy = jest.spyOn(console, 'log')
    await input.setValue('Hola mundo?')
    expect(clsSpy).toHaveBeenCalled()
    expect(getAnswerSpy).toHaveBeenCalled()
  })

  it('getAnswer', async () => {
    const wrapper = shallowMount(Indecision)
    await wrapper.vm.getAnswer()
    expect(wrapper.vm.image).toBe(image)
    expect(wrapper.vm.answer).toBe('Si')
    const input = wrapper.find('input')
    await input.setValue('Hola mundo?')
    const img = wrapper.find('img')
    expect(img.exists()).toBeTruthy()
    expect(img.element.src).toBe(image)
  })

  it('error getAnswer', async () => {
    const wrapper = shallowMount(Indecision)
    fetch.mockImplementationOnce(() => Promise.reject('error'))
    await wrapper.vm.getAnswer()
    expect(wrapper.vm.answer).toBe('error')
  })
})
