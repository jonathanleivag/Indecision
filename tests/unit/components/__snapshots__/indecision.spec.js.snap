// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision component should match the snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda de terminar la pregunta con un signo de interrogaci√≥n (?)</p>
  <!--v-if-->
</div>
`;
